core/ordersync/ordersync.go

```diff
@@ -300,30 +306,47 @@ func (s *Service) GetOrders(ctx context.Context, minPeers int) error {
 			log.WithFields(log.Fields{
 				"provider": peerID.Pretty(),
 			}).Trace("requesting orders from neighbor via ordersync")
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			default:
-			}
 
-			if err := s.getOrdersFromPeer(ctx, peerID); err != nil {
-				log.WithFields(log.Fields{
-					"error":    err.Error(),
-					"provider": peerID.Pretty(),
-				}).Warn("could not get orders from peer via ordersync")
-				continue
-			} else {
-				// TODO(albrow): Handle case where no orders were returned from this
-				// peer. This could be considered a valid response, depending on the implementation
-				// details of the subprotocol. We need to not try them again, but also not count
-				// them toward the number of peers we have successfully synced with.
-				log.WithFields(log.Fields{
-					"provider": peerID.Pretty(),
-				}).Trace("succesfully got orders from peer via ordersync")
-				successfullySyncedPeers.Add(peerID.Pretty())
-			}
+			wg.Add(1)
+			go func(id peer.ID) {
+				select {
+				case <-innerCtx.Done():
+					// NOTE(jalextowle): In this case, we haven't written to the semaphore
+					// so we shouldn't read from it.
+					wg.Done()
+					return
+				case semaphore <- struct{}{}:
+					defer func() {
+						wg.Done()
+						<-semaphore
+					}()
+				}
+				if err := s.getOrdersFromPeer(innerCtx, id); err != nil {
+					log.WithFields(log.Fields{
+						"error":    err.Error(),
+						"provider": id.Pretty(),
+					}).Warn("could not get orders from peer via ordersync")
+				} else {
+					// TODO(albrow): Handle case where no orders were returned from this
+					// peer. This could be considered a valid response, depending on the implementation
+					// details of the subprotocol. We need to not try them again, but also not count
+					// them toward the number of peers we have successfully synced with.
+					log.WithFields(log.Fields{
+						"provider": id.Pretty(),
+					}).Trace("succesfully got orders from peer via ordersync")
+					successfullySyncedPeers.Add(id.Pretty())
```

Comment:
`successfullySyncedPeers` is based on a `map` type under the hood and isn't threadsafe. We need to protect any reads or writes with a mutex.